### PR TITLE
[Feature] Suppport SHOW CREATE ROUTINE LOAD SQL

### PR DESCRIPTION
**Pull request by zaorangyang**
Sometimes users need to know how they created a routine load job, similar to SHOW CREATE TABLE. This PR allows users to obtain the SQL statement for creating the routine load job using SQL.

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 18 / 184 (09.78%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|0|3|00.00%|[45, 46, 48]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|0|43|00.00%|[1541, 1542, 1543, 1544, 1546, 1547, 1549, 1550, 1552, 1553, 1555, 1556, 1558, 1559, 1561, 1562, 1564, 1565, 1567, 1568, 1570, 1571, 1573, 1574, 1576, 1577, 1579, 1580, 1582, 1583, 1585, 1586, 1587, 1590, 1591, 1593, 1594, 1596, 1597, 1599, 1600, 1602, 1603]|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|[347]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|0|16|00.00%|[24, 25, 26, 27, 28, 33, 34, 37, 38, 39, 43, 47, 51, 52, 56, 61]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|49|02.04%|[296, 1389, 1390, 1391, 1392, 1393, 1396, 1397, 1398, 1399, 1400, 1401, 1402, 1404, 1408, 1409, 1412, 1413, 1416, 1417, 1418, 1420, 1421, 1422, 1423, 1424, 1425, 1426, 1428, 1429, 1430, 1432, 1434, 1435, 1436, 1437, 1438, 1439, 1441, 1442, 1443, 1446, 1447, 1448, 1449, 1450, 1451, 1452]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|17|71|23.94%|[528, 536, 537, 538, 539, 541, 542, 544, 545, 546, 547, 548, 549, 550, 553, 556, 557, 558, 559, 560, 561, 562, 563, 565, 566, 567, 568, 569, 570, 572, 574, 575, 578, 580, 581, 582, 585, 586, 587, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 601, 602, 619, 632]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 185 (17.84%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|0|3|00.00%|[45, 46, 48]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|0|43|00.00%|[1541, 1542, 1543, 1544, 1546, 1547, 1549, 1550, 1552, 1553, 1555, 1556, 1558, 1559, 1561, 1562, 1564, 1565, 1567, 1568, 1570, 1571, 1573, 1574, 1576, 1577, 1579, 1580, 1582, 1583, 1585, 1586, 1587, 1590, 1591, 1593, 1594, 1596, 1597, 1599, 1600, 1602, 1603]|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|50|02.00%|[296, 1389, 1390, 1391, 1392, 1393, 1396, 1397, 1398, 1399, 1400, 1401, 1402, 1404, 1408, 1409, 1412, 1413, 1416, 1417, 1418, 1420, 1421, 1422, 1423, 1424, 1425, 1426, 1428, 1429, 1430, 1432, 1434, 1435, 1436, 1437, 1438, 1439, 1441, 1442, 1443, 1446, 1447, 1448, 1449, 1450, 1451, 1452, 1453]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|17|71|23.94%|[528, 536, 537, 538, 539, 541, 542, 544, 545, 546, 547, 548, 549, 550, 553, 556, 557, 558, 559, 560, 561, 562, 563, 565, 566, 567, 568, 569, 570, 572, 574, 575, 578, 580, 581, 582, 585, 586, 587, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 601, 602, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 186 (17.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|0|3|00.00%|[45, 46, 48]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|0|43|00.00%|[1541, 1542, 1543, 1544, 1546, 1547, 1549, 1550, 1552, 1553, 1555, 1556, 1558, 1559, 1561, 1562, 1564, 1565, 1567, 1568, 1570, 1571, 1573, 1574, 1576, 1577, 1579, 1580, 1582, 1583, 1585, 1586, 1587, 1590, 1591, 1593, 1594, 1596, 1597, 1599, 1600, 1602, 1603]|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|51|01.96%|[296, 1389, 1390, 1391, 1392, 1393, 1396, 1397, 1398, 1399, 1400, 1401, 1402, 1404, 1408, 1409, 1412, 1413, 1416, 1417, 1418, 1420, 1421, 1422, 1423, 1424, 1425, 1426, 1428, 1429, 1430, 1432, 1434, 1435, 1436, 1437, 1438, 1439, 1440, 1441, 1443, 1444, 1445, 1448, 1449, 1450, 1451, 1452, 1453, 1454]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|17|71|23.94%|[528, 536, 537, 538, 539, 541, 542, 544, 545, 546, 547, 548, 549, 550, 553, 556, 557, 558, 559, 560, 561, 562, 563, 565, 566, 567, 568, 569, 570, 572, 574, 575, 578, 580, 581, 582, 585, 586, 587, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 601, 602, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 33 / 186 (17.74%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|0|3|00.00%|[45, 46, 48]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|0|43|00.00%|[1541, 1542, 1543, 1544, 1546, 1547, 1549, 1550, 1552, 1553, 1555, 1556, 1558, 1559, 1561, 1562, 1564, 1565, 1567, 1568, 1570, 1571, 1573, 1574, 1576, 1577, 1579, 1580, 1582, 1583, 1585, 1586, 1587, 1590, 1591, 1593, 1594, 1596, 1597, 1599, 1600, 1602, 1603]|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|1|51|01.96%|[296, 1389, 1390, 1391, 1392, 1393, 1396, 1397, 1398, 1399, 1400, 1401, 1402, 1404, 1408, 1409, 1412, 1413, 1416, 1417, 1418, 1420, 1421, 1422, 1423, 1424, 1425, 1426, 1428, 1429, 1430, 1432, 1434, 1435, 1436, 1437, 1438, 1439, 1440, 1441, 1443, 1444, 1445, 1448, 1449, 1450, 1451, 1452, 1453, 1454]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|17|71|23.94%|[528, 536, 537, 538, 539, 541, 542, 544, 545, 546, 547, 548, 549, 550, 553, 556, 557, 558, 559, 560, 561, 562, 563, 565, 566, 567, 568, 569, 570, 572, 574, 575, 578, 580, 581, 582, 585, 586, 587, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 601, 602, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 127 / 184 (69.02%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|16|49|32.65%|[1390, 1391, 1392, 1393, 1394, 1395, 1397, 1399, 1401, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1412, 1413, 1414, 1415, 1416, 1417, 1418, 1420, 1424, 1425, 1436, 1437, 1438, 1441, 1442, 1446, 1454]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|2|3|66.67%|[46]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|53|71|74.65%|[528, 546, 547, 548, 549, 553, 569, 570, 572, 575, 591, 592, 593, 594, 595, 596, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|41|43|95.35%|[1586, 1587]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 34 / 184 (18.48%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|0|3|00.00%|[45, 46, 48]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|0|43|00.00%|[1541, 1542, 1543, 1544, 1546, 1547, 1549, 1550, 1552, 1553, 1555, 1556, 1558, 1559, 1561, 1562, 1564, 1565, 1567, 1568, 1570, 1571, 1573, 1574, 1576, 1577, 1579, 1580, 1582, 1583, 1585, 1586, 1587, 1590, 1591, 1593, 1594, 1596, 1597, 1599, 1600, 1602, 1603]|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|2|49|04.08%|[1390, 1391, 1392, 1393, 1394, 1395, 1397, 1399, 1401, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1412, 1413, 1414, 1415, 1416, 1417, 1418, 1420, 1424, 1425, 1428, 1429, 1432, 1433, 1434, 1436, 1437, 1438, 1439, 1440, 1441, 1442, 1444, 1445, 1446, 1448, 1450, 1451, 1452, 1453, 1454]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|17|71|23.94%|[528, 536, 537, 538, 539, 541, 542, 544, 545, 546, 547, 548, 549, 550, 553, 556, 557, 558, 559, 560, 561, 562, 563, 565, 566, 567, 568, 569, 570, 572, 574, 575, 578, 580, 581, 582, 585, 586, 587, 589, 590, 591, 592, 593, 594, 595, 596, 597, 598, 599, 601, 602, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 127 / 184 (69.02%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|16|49|32.65%|[1390, 1391, 1392, 1393, 1394, 1395, 1397, 1399, 1401, 1402, 1403, 1405, 1406, 1407, 1408, 1409, 1412, 1413, 1414, 1415, 1416, 1417, 1418, 1420, 1424, 1425, 1436, 1437, 1438, 1441, 1442, 1446, 1454]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|2|3|66.67%|[46]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|53|71|74.65%|[528, 546, 547, 548, 549, 553, 569, 570, 572, 575, 591, 592, 593, 594, 595, 596, 619, 632]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|41|43|95.35%|[1586, 1587]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Comment by meegoo** [comment]
pls add some description about this PR

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java [hunk 1]
```diff
@@ -1381,6 +1385,73 @@ private void handleShowLoad() throws AnalysisException {
...
+        if (routineLoadJob.getPartitions() != null) {
+            createRoutineLoadSql.append(",\n");
+            createRoutineLoadSql.append(routineLoadJob.getPartitions().toString());
+        }
+        if (routineLoadJob.getWhereExpr() != null) {
+            createRoutineLoadSql.append(",\nWHERE ");
+            createRoutineLoadSql.append(routineLoadJob.getWhereExpr().toSql());
+        }
+
+        createRoutineLoadSql.append("\nPROPERTIES\n").append(routineLoadJob.jobPropertiesToSql());
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:1447]: If there are some secret message in the property, how to show them?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 2]
```diff
@@ -1077,6 +1078,10 @@ showRoutineLoadTaskStatement
         WHERE expression
     ;
 
+showCreateRoutineLoadStatement
+    : SHOW CREATE ROUTINE LOAD FOR (db=qualifiedName '.')? name=identifier
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1082]: Like `show create table xxx` , maybe the `FOR` is not necessary.

File: fe/fe-core/src/main/java/com/starrocks/load/routineload/KafkaRoutineLoadJob.java [hunk 3]
```diff
@@ -517,17 +525,84 @@ private void setCustomKafkaProperties(Map<String, String> kafkaProperties) {
...
     @Override
-    protected String dataSourcePropertiesJsonToString() {
-        Map<String, String> dataSourceProperties = Maps.newHashMap();
-        dataSourceProperties.put("brokerList", brokerList);
-        dataSourceProperties.put("topic", topic);
+    public String dataSourcePropertiesToSql() {
+        StringBuilder sb = new StringBuilder();
+        sb.append("(\n");
+        sb.append("\"").append(CreateRoutineLoadStmt.KAFKA_BROKER_LIST_PROPERTY).append("\"=\"");
+        sb.append(brokerList).append("\",\n");
```
Comment (rickif) [inline fe/fe-core/src/main/java/com/starrocks/load/routineload/KafkaRoutineLoadJob.java:539]: Is it necessarry to add `\n` here?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 2 above)
Comment (zaorangyang) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1082]: > Like `show create table xxx` , maybe the `FOR` is not necessary.

@jaogoy What do you think?

File: fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java (hunk 1 above)
Comment (zaorangyang) [reply fe/fe-core/src/main/java/com/starrocks/qe/ShowExecutor.java:1447]: @chaoyli 
image
Like this.

**Comment by zaorangyang** [comment]
@meegoo updated

**Comment by zaorangyang** [comment]
> Could you give an example of the result of `SHOW CREATE ROUTINE LOAD`?

@rickif 
```
CREATE ROUTINE LOAD herb.testherbavro225 on herbavro
 COLUMNS TERMINATED BY '\t',
COLUMNS (advertiser_id, advertiser_ad_type, advertiser_ad_width, advertiser_currency_type),
PARTITIONS (p1, p2, p3),
WHERE `advertiser_id` > -100
PROPERTIES
(
"desired_concurrent_number"="1",
"max_error_number"="1000",
"max_filter_ratio"="1.0",
"max_batch_interval"="5",
"max_batch_rows"="200000",
"task_consume_second"="15",
"task_timeout_second"="60",
"format"="avro",
"jsonpaths"="",
"strip_outer_array"="false",
"json_root"="",
"strict_mode"="false",
"timezone"="Asia/Shanghai",
"partial_update"="false",
"trim_space"="false",
"enclose"="0",
"escape"="0",
"log_rejected_record_num"="0"
)
FROM KAFKA
(
"kafka_broker_list"="172.26.194.239:9092",
"kafka_topic"="herb-avro",
"kafka_partitions"="0",
"kafka_offsets"="OFFSET_BEGINNING",
"confluent.schema.registry.url"="http://172.26.194.240:8081"
);
```

File: fe/fe-core/src/main/java/com/starrocks/load/routineload/KafkaRoutineLoadJob.java (hunk 3 above)
Comment (zaorangyang) [reply fe/fe-core/src/main/java/com/starrocks/load/routineload/KafkaRoutineLoadJob.java:539]: > Is it necessarry to add `\n` here?

Yes, each attribute should be on a new line.

**Review by rickif — APPROVED** [review]

**Review by rickif — APPROVED** [review]

**Review by meegoo — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 153 / 186 (82.26%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/load/routineload/PulsarRoutineLoadJob.java|0|1|00.00%|[279]|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnSeparator.java|2|3|66.67%|[46]|
|:large_blue_circle:|com/starrocks/load/routineload/KafkaRoutineLoadJob.java|53|71|74.65%|[528, 546, 547, 548, 549, 553, 569, 570, 572, 575, 591, 592, 593, 594, 595, 596, 619, 632]|
|:large_blue_circle:|com/starrocks/qe/ShowExecutor.java|42|51|82.35%|[1460, 1461, 1462, 1465, 1466, 1470, 1478, 1479, 1480]|
|:large_blue_circle:|com/starrocks/sql/ast/ShowCreateRoutineLoadStmt.java|14|16|87.50%|[50, 51]|
|:large_blue_circle:|com/starrocks/load/routineload/RoutineLoadJob.java|41|43|95.35%|[1611, 1612]|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|[]|

**Review by chaoyli — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Comment by zaorangyang** [comment]
@Mergify backport branch-3.1

**Pull request merged**